File: java/blackjack/src/main/java/Blackjack.java
import java.util.HashMap;
import java.util.Map;

public class Blackjack {

    private static final int SCORE = 21;
    private static Map<String, Integer> cardValues;

    static {
        cardValues = new HashMap<>();
        cardValues.put("ace", 11);
        cardValues.put("two", 2);
        cardValues.put("three", 3);
        cardValues.put("four", 4);
        cardValues.put("five", 5);
        cardValues.put("six", 6);
        cardValues.put("seven", 7);
        cardValues.put("eight", 8);
        cardValues.put("nine", 9);
        cardValues.put("ten", 10);
        cardValues.put("jack", 10);
        cardValues.put("queen", 10);
        cardValues.put("king", 10);
        cardValues.put("other", 0);
    }

    public int parseCard(String card) {
        return cardValues.getOrDefault(card, 0);
    }

    public boolean isBlackjack(String card1, String card2) {

        var value = parseCard(card1) + parseCard(card2);

        return value == SCORE;
    }

    public String largeHand(boolean isBlackjack, int dealerScore) {

        if (!isBlackjack) {
            return "P";
        }

        if (isBlackjack && dealerScore >= 10) {
            return "S";
        } else {
            return "W";
        }
    }

    public String smallHand(int handScore, int dealerScore) {

        // This is a bit ugly
        if (handScore >= 17) {
            return "S";
        } else if (handScore <= 16 && handScore >= 12) {

            if (dealerScore >= 7) {
                return "H";
            } else {
                return "S";
            }
        } else  {
            // handScore <= 11
            return "H";
        }
    }


    // FirstTurn returns the semi-optimal decision for the first turn, given the cards of the player and the dealer.
    // This function is already implemented and does not need to be edited. It pulls the other functions together in a
    // complete decision tree for the first turn.
    public String firstTurn(String card1, String card2, String dealerCard) {

        int handScore = parseCard(card1) + parseCard(card2);
        int dealerScore = parseCard(dealerCard);

        if (20 < handScore) {
            return largeHand(isBlackjack(card1, card2), dealerScore);
        } else {
            return smallHand(handScore, dealerScore);
        }
    }
}

File: go/blackjack/blackjack.go
package blackjack

var cardValue = map[string]int{
	"ace":   11,
	"jack":  10,
	"king":  10,
	"queen": 10,
	"ten":   10,
	"nine":  9,
	"eight": 8,
	"seven": 7,
	"six":   6,
	"five":  5,
	"four":  4,
	"three": 3,
	"two":   2,
	"one":   1,
}

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {

	if value, ok := cardValue[card]; ok {
		return value
	} else {
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {

	cardSum := ParseCard(card1) + ParseCard(card2)
	dealerValue := ParseCard(dealerCard)

	// these just seem like random placements
	// logic could be cleaned up
	// just directly translated text instructions to code
	switch {
	case card1 == "ace" && card2 == "ace":
		return "P"
	case cardSum == 21 && (dealerValue < 10):
		return "W"
	case cardSum == 21 && (dealerValue >= 10):
		return "S"
	case cardSum >= 17 && cardSum <= 20:
		return "S"
	case cardSum >= 12 && cardSum <= 16 && ParseCard(dealerCard) < 7:
		return "S"
	case cardSum == 11:
		return "H"
	default:
		// hit as the default
		return "H"
	}
}
